File: projects/composablekernel/python/ck_dsl_c/tests/parity/fmha_appendkv_emit.c
/* Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
 * SPDX-License-Identifier: MIT
 *
 * tests/parity/fmha_appendkv_emit.c -- C-side emitter for the fmha_appendkv
 * parity harness. Selects one of 6 FmhaAppendKvSpec configs by argv[1] (0..5),
 * builds ckc_fmha_appendkv_spec_t identically to the Python emitter
 * fmha_appendkv_emit.py, builds the kernel via ckc_build_fmha_fwd_appendkv and
 * lowers via ckc_lower_kernel_to_llvm_ex (arch gfx950, flavor AUTO), printing
 * the .ll to stdout so the two outputs can be byte-compared.
 */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "ckc/ir.h"
#include "ckc/lower_llvm.h"
#include "ckc/instance_fmha_appendkv.h"
#include "ckc/helper_ck_dsl.instances.common._fmha_common.h"
#include "ckc/helper_ck_dsl.helpers.rotary.h"

/* Fill `spec` for config index `idx`. Returns 0 on success, -1 if unknown. */
static int make_spec(int idx, ckc_fmha_appendkv_spec_t *spec) {
    ckc_fmha_shape_t shape;
    ckc_fmha_common_spec_t common;
    ckc_rotary_spec_t rot;

    switch (idx) {
    case 0: { /* H128 q4 kv2 f16, batch1, no rope, b256 */
        shape = ckc_fmha_shape_default(128, 4, 2);
        common = ckc_fmha_common_spec_default(shape);
        common.dtype = "f16";
        *spec = ckc_fmha_appendkv_spec_default(common, 1);
        spec->has_rotary = false;
        spec->block_size = 256;
        break;
    }
    case 1: { /* H128 q4 kv2 f16, batch2, rope half H128, b256 */
        shape = ckc_fmha_shape_default(128, 4, 2);
        common = ckc_fmha_common_spec_default(shape);
        common.dtype = "f16";
        *spec = ckc_fmha_appendkv_spec_default(common, 2);
        if (ckc_rotary_spec_init(&rot, 128, CKC_ROTARY_HALF, 0) != CKC_OK)
            return -1;
        spec->has_rotary = true;
        spec->rotary = rot;
        spec->block_size = 256;
        break;
    }
    case 2: { /* H128 q8 kv4 bsq16 bsk64 bf16, batch2, rope interleaved H128, b128 */
        shape = ckc_fmha_shape_make(128, 8, 4, 16, 64);
        common = ckc_fmha_common_spec_default(shape);
        common.dtype = "bf16";
        *spec = ckc_fmha_appendkv_spec_default(common, 2);
        if (ckc_rotary_spec_init(&rot, 128, CKC_ROTARY_INTERLEAVED, 0) != CKC_OK)
            return -1;
        spec->has_rotary = true;
        spec->rotary = rot;
        spec->block_size = 128;
        break;
    }
    case 3: { /* H64 q4 kv2 f16, batch1, rope half H64, b256 */
        shape = ckc_fmha_shape_default(64, 4, 2);
        common = ckc_fmha_common_spec_default(shape);
        common.dtype = "f16";
        *spec = ckc_fmha_appendkv_spec_default(common, 1);
        if (ckc_rotary_spec_init(&rot, 64, CKC_ROTARY_HALF, 0) != CKC_OK)
            return -1;
        spec->has_rotary = true;
        spec->rotary = rot;
        spec->block_size = 256;
        break;
    }
    case 4: { /* H32 q8 kv8 bf16, batch2, no rope, b256 */
        shape = ckc_fmha_shape_default(32, 8, 8);
        common = ckc_fmha_common_spec_default(shape);
        common.dtype = "bf16";
        *spec = ckc_fmha_appendkv_spec_default(common, 2);
        spec->has_rotary = false;
        spec->block_size = 256;
        break;
    }
    case 5: { /* H192 q4 kv2 f16, batch1, rope half H192, b256 */
        shape = ckc_fmha_shape_default(192, 4, 2);
        common = ckc_fmha_common_spec_default(shape);
        common.dtype = "f16";
        *spec = ckc_fmha_appendkv_spec_default(common, 1);
        if (ckc_rotary_spec_init(&rot, 192, CKC_ROTARY_HALF, 0) != CKC_OK)
            return -1;
        spec->has_rotary = true;
        spec->rotary = rot;
        spec->block_size = 256;
        break;
    }
    default:
        return -1;
    }
    return 0;
}

int main(int argc, char **argv) {
    if (argc < 2) {
        fprintf(stderr, "usage: %s <config_index 0..5>\n", argv[0]);
        return 2;
    }
    int idx = atoi(argv[1]);

    ckc_fmha_appendkv_spec_t spec;
    if (make_spec(idx, &spec) != 0) {
        fprintf(stderr, "unknown/invalid config index %d\n", idx);
        return 2;
    }

    /* Build via the requested entry: ckc_build_fmha_fwd_appendkv(b, spec, arch),
     * with the builder initialised to spec.kernel_name() (the convenience
     * wrapper does exactly this). */
    ckc_ir_builder_t b;
    char namebuf[256];
    if (ckc_fmha_appendkv_kernel_name(&spec, namebuf, sizeof namebuf) != CKC_OK) {
        fprintf(stderr, "kernel_name failed\n");
        return 1;
    }
    if (ckc_ir_builder_init(&b, namebuf) != CKC_OK) {
        fprintf(stderr, "builder init failed\n");
        return 1;
    }

    ckc_kernel_def_t *kernel = ckc_build_fmha_fwd_appendkv(&b, &spec, "gfx950");
    if (kernel == NULL) {
        fprintf(stderr, "build failed: %s\n", ckc_ir_builder_error(&b));
        ckc_ir_builder_free(&b);
        return 1;
    }

    char *llvm_text = NULL;
    char err[CKC_ERR_MSG_CAP];
    err[0] = 0;
    ckc_status_t st = ckc_lower_kernel_to_llvm_ex(kernel, CKC_LLVM_FLAVOR_AUTO,
                                                  "gfx950", &llvm_text,
                                                  err, sizeof err);
    if (st != CKC_OK || !llvm_text) {
        fprintf(stderr, "lower failed: status=%d err=%s\n", (int)st, err);
        ckc_ir_builder_free(&b);
        return 1;
    }
    fputs(llvm_text, stdout);
    free(llvm_text);
    ckc_ir_builder_free(&b);
    return 0;
}

File: projects/composablekernel/python/ck_dsl_c/tests/parity/fmha_appendkv_emit.py
#!/usr/bin/env python3
# Copyright (c) Advanced Micro Devices, Inc., or its affiliates.
# SPDX-License-Identifier: MIT
#
# tests/parity/fmha_appendkv_emit.py -- Python reference emitter for the
# fmha_appendkv parity harness. Selects one of 6 FmhaAppendKvSpec configs by
# argv[1] (0..5), builds via build_fmha_fwd_appendkv and prints
# lower_kernel_to_llvm(arch='gfx950') so it can be byte-compared with the C
# emitter fmha_appendkv_emit.c.
import sys

from ck_dsl.instances.common._fmha_common import FmhaCommonSpec, FmhaShape
from ck_dsl.instances.common.fmha_appendkv import (
    FmhaAppendKvSpec,
    build_fmha_fwd_appendkv,
)
from ck_dsl.helpers.rotary import RotarySpec
from ck_dsl import lower_kernel_to_llvm
from ck_dsl.core.ir_serialize import serialize
from ck_dsl.core.verify import verify


def _spec(idx: int) -> FmhaAppendKvSpec:
    if idx == 0:
        return FmhaAppendKvSpec(
            common=FmhaCommonSpec(
                shape=FmhaShape(head_size=128, num_query_heads=4, num_kv_heads=2),
                dtype="f16",
            ),
            batch=1,
            rotary=None,
            block_size=256,
        )
    if idx == 1:
        return FmhaAppendKvSpec(
            common=FmhaCommonSpec(
                shape=FmhaShape(head_size=128, num_query_heads=4, num_kv_heads=2),
                dtype="f16",
            ),
            batch=2,
            rotary=RotarySpec(head_size=128, layout="half"),
            block_size=256,
        )
    if idx == 2:
        return FmhaAppendKvSpec(
            common=FmhaCommonSpec(
                shape=FmhaShape(
                    head_size=128,
                    num_query_heads=8,
                    num_kv_heads=4,
                    block_size_q=16,
                    block_size_k=64,
                ),
                dtype="bf16",
            ),
            batch=2,
            rotary=RotarySpec(head_size=128, layout="interleaved"),
            block_size=128,
        )
    if idx == 3:
        return FmhaAppendKvSpec(
            common=FmhaCommonSpec(
                shape=FmhaShape(head_size=64, num_query_heads=4, num_kv_heads=2),
                dtype="f16",
            ),
            batch=1,
            rotary=RotarySpec(head_size=64, layout="half"),
            block_size=256,
        )
    if idx == 4:
        return FmhaAppendKvSpec(
            common=FmhaCommonSpec(
                shape=FmhaShape(head_size=32, num_query_heads=8, num_kv_heads=8),
                dtype="bf16",
            ),
            batch=2,
            rotary=None,
            block_size=256,
        )
    if idx == 5:
        return FmhaAppendKvSpec(
            common=FmhaCommonSpec(
                shape=FmhaShape(head_size=192, num_query_heads=4, num_kv_heads=2),
                dtype="f16",
            ),
            batch=1,
            rotary=RotarySpec(head_size=192, layout="half"),
            block_size=256,
        )
    raise SystemExit(f"unknown config index {idx}")


def main() -> int:
    if len(sys.argv) < 2:
        sys.stderr.write("usage: fmha_appendkv_emit.py <config_index 0..5>\n")
        return 2
    idx = int(sys.argv[1])
    mode = sys.argv[2] if len(sys.argv) > 2 else "ll"
    spec = _spec(idx)
    kernel = build_fmha_fwd_appendkv(spec, arch="gfx950")
    if mode == "ll":
        text = lower_kernel_to_llvm(kernel, arch="gfx950")
        sys.stdout.write(text)
    elif mode == "ir":
        sys.stdout.write(serialize(kernel))
    elif mode == "verify":
        sys.stdout.write("".join(str(d) + "\n" for d in verify(kernel)))
    else:
        sys.stderr.write(f"unknown mode {mode}\n")
        return 2
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
